src/diagnostic-channel-publishers/src/pg.pub.ts

```diff
@@ -0,0 +1,164 @@
+                const end = process.hrtime(start);
+                data.result = res && { rowCount: res.rowCount, command: res.command };
+                data.error = err;
+                data.duration = Math.ceil((end[0] * 1e3) + (end[1] / 1e6));
+                channel.publish("postgres", data);
+
+                // emulate weird internal behavior in pg@6
+                // on success, the callback is called *before* query events are emitted
+                // on failure, the callback is called *instead of* the query emitting events
+                // with no events, that means no promises (since the promise is resolved/rejected in an event handler)
+                // since we are always inserting ourselves as a callback, we have to restore the original
+                // behavior if the user didn't provide one themselves
+                if (err) {
+                    if (cb) {
+                        return cb.apply(this, arguments);
+                    } else if (queryResult && queryResult instanceof EventEmitter) {
+                        queryResult.emit("error", err);
+                    }
+                } else if (cb) {
+                    cb.apply(this, arguments);
+                }
+            });
+
+            if (cachePatched) {
+                // this should only happen in the case of config.callback = patch(config.callback)
+                patchedCallbacks.set(trackingCallback, trackingCallback);
+            }
+
+            if (cb) {
+                patchedCallbacks.set(cb, trackingCallback);
+            }
+
+            return trackingCallback;
+        }
+
+        // this function takes too many variations of aguments.
+        // this patches any provided callback or creates a new callback if one wasn't provided.
+        // since the callback is always called (if provided) in addition to always having a Promisified
+        // EventEmitter returned (well, sometimes -- see above), its safe to insert a callback if none was given
+        if (typeof config === "string") {
+            if (values instanceof Array) {
+                data.query.preparable = {
+                    text: config,
+                    args: values,
+                };
+                callback = patchCallback(callback);
+            } else {
+                data.query.text = config;
+
+                // pg v6 will, for some reason, accept both
+                // client.query("...", undefined, () => {...})
+                // **and**
+                // client.query("...", () => {...});
+                // Internally, precedence is given to the callback argument
+                if (callback) {
+                    callback = patchCallback(callback);
+                } else {
+                    values = patchCallback(values);
+                }
+            }
+        } else {
+            if (typeof config.name === "string") {
```

Comment:
So this is all based on that normalizeQueryArgs loosely, but I didn't want to do what that function does and create my own config object because I wanted to pass in the args as closely to the original call as possible (aside from possibly adding a callback to the end of the args list if need be).